### PR TITLE
DC/RC: added callbacks into handling failures

### PR DESCRIPTION
**Pull request by hoopoepg**
- to provide DCI specific failure handling added ops
  to DCI structore
- added callback to rx_purge_outstanding function to allow
  custom purge algorithms

### CODE REVIEW

File: src/uct/ib/rc/base/rc_ep.h [hunk 1]
```diff
@@ -248,8 +254,14 @@ void uct_rc_fc_cleanup(uct_rc_fc_t *fc);
 
 ucs_status_t uct_rc_ep_fc_grant(uct_pending_req_t *self);
 
+void uct_rc_txqp_purge_outstanding_cb(uct_rc_iface_t *iface, uct_rc_txqp_t *txqp,
```
Comment (brminich) [inline src/uct/ib/rc/base/rc_ep.h:257]: why need to change this semantic?

Comment (hoopoepg) [reply src/uct/ib/rc/base/rc_ep.h:257]: because DCX failure may require own logic: re-post requests 2-N

Comment (brminich) [reply src/uct/ib/rc/base/rc_ep.h:257]: can't you just invoke different func for keepalive DCI?

Comment (hoopoepg) [reply src/uct/ib/rc/base/rc_ep.h:257]: I can, and in arbiter as well, but it will produce code duplication

Comment (hoopoepg) [reply src/uct/ib/rc/base/rc_ep.h]: reverted purge_outstanding changes

**Review by brminich — COMMENTED** [review]
👍, beside minor comment

**Comment by hoopoepg** [comment]
bot:pipe:retest

File: src/uct/ib/dc/dc_mlx5.c [hunk 2]
```diff
@@ -777,14 +777,17 @@ void uct_dc_mlx5_iface_dcis_destroy(uct_dc_mlx5_iface_t *iface, int max)
...
 {
     struct ibv_qp_cap cap = {};
 
+    dci->ops = ops;
     return uct_dc_mlx5_iface_create_qp(iface, &cap, dci);
 }
 
-static ucs_status_t uct_dc_mlx5_iface_create_dcis(uct_dc_mlx5_iface_t *iface)
+static ucs_status_t uct_dc_mlx5_iface_create_dcis(uct_dc_mlx5_iface_t *iface,
+                                                  uct_dc_dci_ops_t *ops)
```
[the remark this answers is not in the document]
Comment (hoopoepg) [reply src/uct/ib/dc/dc_mlx5.c]: done

**Review by brminich — APPROVED** [review]

File: src/uct/ib/dc/dc_mlx5.h [hunk 3]
```diff
@@ -113,7 +113,19 @@ typedef struct uct_dc_mlx5_iface_config {
...
 
+typedef void (*uct_dc_dci_handle_failure_func_t)(uct_dc_mlx5_iface_t *iface,
+                                                 struct mlx5_cqe64 *cqe,
+                                                 uint8_t dci,
+                                                 ucs_status_t status);
+
+
+typedef struct uct_dc_dci_ops {
+    uct_dc_dci_handle_failure_func_t handle_failure;
+} uct_dc_dci_ops_t;
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5.h:124]: pls add doc

Comment (hoopoepg) [reply src/uct/ib/dc/dc_mlx5.h]: added comments

File: src/uct/ib/dc/dc_mlx5.h [hunk 4]
```diff
@@ -119,8 +119,11 @@ typedef void (*uct_dc_dci_handle_failure_func_t)(uct_dc_mlx5_iface_t *iface,
                                                  ucs_status_t status);
 
 
+/**
+ * DCI interface operations.
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5.h:123]: DCI QP-specific operations

File: src/uct/ib/dc/dc_mlx5.h [hunk 5]
```diff
@@ -119,8 +119,11 @@ typedef void (*uct_dc_dci_handle_failure_func_t)(uct_dc_mlx5_iface_t *iface,
...
+ */
 typedef struct uct_dc_dci_ops {
-    uct_dc_dci_handle_failure_func_t handle_failure;
+    uct_dc_dci_handle_failure_func_t handle_failure; /* error handling */
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5.h:126]: callback for handling completion with error on the DCI

File: src/uct/ib/dc/dc_mlx5.h (hunk 4 above)
Comment (hoopoepg) [reply src/uct/ib/dc/dc_mlx5.h]: reworded

File: src/uct/ib/dc/dc_mlx5.h (hunk 5 above)
Comment (hoopoepg) [reply src/uct/ib/dc/dc_mlx5.h]: reworded

**Review by yosefe — APPROVED** [review]

**Comment by hoopoepg** [comment]
squashed
have to rebase due to merge conflicts

**Comment by hoopoepg** [comment]
bot:pipe:retest

**Review by yosefe — APPROVED** [review]

**Comment by yosefe** [comment]
bot:pipe:retest

**Comment by hoopoepg** [comment]
timeout

**Comment by hoopoepg** [comment]
bot:pipe:retest

**Pull request merged**